Azure/sonic-buildimage
# [PINS] Add genl_packet kernel module for PINS packet ins

- Add genl_packet kernel module
- Extend bcm-knet-cb to support genl_packet packets
- Include kernel module in barefoot and broadcom images
- Update copp_cfg file with packet in queue configuration

Submission containing materials of a third party:
    Copyright Google LLC; Licensed under Apache 2.0 and GPL v2

Co-authored-by: Vivek Ramamoorthy <vivekmoorthy@google.com>
Co-authored-by: Srikishen Pondicherry Shanmugam <kishanps@google.com>
Co-authored-by: Yilan Ji <yilanji@google.com>
Co-authored-by: Brian O'Connor <bocon@opennetworking.org>

#### Why I did it

Adds support for receiving packet ins with additonal metadata for PINS.

See this HLD for more details:
https://github.com/pins/SONiC/blob/master/doc/pins/Packet_io.md

#### How I did it

This code largely follows the same pattern as psample for sflow.

#### How to verify it

Build a broadcom or barefoot image with INCLUDE_P4RT=y

Install P4RT flow rules for LLDP.

Send LLDP packet from peer using bcmsh:
```
tx 1 PBM=cd0 DATA="01 80 c2 00 00 0e...
```

Verify counters and debug logs(dmesg) for knet-cb:
```
cat /proc/bcm/knet-cb/genl_packet/stats
BCM KNET generic Callback Stats
  DCB type 38
  pkts filter generic cb                  1
  pkts sent to generic module             1
  pkts handled by generic                 1
  pkts with mc destination                0
  pkts with cpu destination               1
  pkts current queue length               0
  pkts high queue length                  1
  pkts drop max queue length              0
  pkts drop no memory                     0
  pkts drop generic not ready             0
  pkts drop metadata parse error          0
  pkts with invalid src port              0
  pkts with invalid dst port              0
  pkts with invalid orig pkt sz           0
```

#### Which release branch to backport (provide reason below if selected)

None

#### Description for the changelog

Add genl_packet kernel module for PINS packet ins
